### PR TITLE
Migrate to from Nom 7 to Winnow 0.3

### PR DESCRIPTION
**Pull request by epage**
Benefits over Nom
- Maintained
- Ergonomics ([1](https://epage.github.io/blog/2023/02/winnow-toml-edit-combine-nom/#winnow), [2](https://epage.github.io/blog/2023/07/winnow-0-5-the-fastest-rust-parser-combinator-library/))
- Performance ([3](https://epage.github.io/blog/2023/07/winnow-0-5-the-fastest-rust-parser-combinator-library/#numbers))

Winnow is at 0.6 but the 0.4->0.5 upgrade is rough enough (and this is a low enough priority) that I've been taking a while on this which has led to dealing with too many merge conflicts.  I'm starting off with just Winnow 0.3 as this last round of merge conflicts already had me rewrite a large number of commits that I want something in sooner than later.

There is a decent chance that future upgrades will improve performance further

<details>

<summary>Performance:</summary>

```
     Running benches/from_str.rs (/home/epage/src/personal/rinja/target/release/deps/from_str-453ae6795003ee85)
Gnuplot not found, using plotters backend
librustdoc/all          time:   [183.26 µs 183.71 µs 184.18 µs]
                        thrpt:  [76.669 MiB/s 76.866 MiB/s 77.053 MiB/s]
                 change:
                        time:   [-16.432% -15.530% -14.572%] (p = 0.00 < 0.05)
                        thrpt:  [+17.057% +18.386% +19.662%]
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
librustdoc/item_info    time:   [3.1356 µs 3.1388 µs 3.1415 µs]
                        thrpt:  [50.089 MiB/s 50.133 MiB/s 50.185 MiB/s]
                 change:
                        time:   [-17.301% -16.994% -16.637%] (p = 0.00 < 0.05)
                        thrpt:  [+19.958% +20.473% +20.921%]
                        Performance has improved.
Found 14 outliers among 100 measurements (14.00%)
  6 (6.00%) low severe
  3 (3.00%) low mild
  5 (5.00%) high severe
librustdoc/item_union   time:   [18.955 µs 19.113 µs 19.301 µs]
                        thrpt:  [51.140 MiB/s 51.642 MiB/s 52.073 MiB/s]
                 change:
                        time:   [-14.048% -13.083% -11.600%] (p = 0.00 < 0.05)
                        thrpt:  [+13.122% +15.052% +16.344%]
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) low mild
  3 (3.00%) high severe
librustdoc/page         time:   [91.669 µs 91.894 µs 92.136 µs]
                        thrpt:  [67.208 MiB/s 67.384 MiB/s 67.549 MiB/s]
                 change:
                        time:   [-15.574% -15.175% -14.731%] (p = 0.00 < 0.05)
                        thrpt:  [+17.276% +17.889% +18.448%]
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
librustdoc/print_item   time:   [10.168 µs 10.187 µs 10.205 µs]
                        thrpt:  [92.521 MiB/s 92.676 MiB/s 92.855 MiB/s]
                 change:
                        time:   [-17.705% -17.536% -17.363%] (p = 0.00 < 0.05)
                        thrpt:  [+21.011% +21.264% +21.514%]
                        Performance has improved.
Found 18 outliers among 100 measurements (18.00%)
  7 (7.00%) low severe
  7 (7.00%) low mild
  3 (3.00%) high mild
  1 (1.00%) high severe
librustdoc/short_item_info
                        time:   [10.968 µs 10.985 µs 11.008 µs]
                        thrpt:  [82.305 MiB/s 82.477 MiB/s 82.606 MiB/s]
                 change:
                        time:   [-11.565% -11.330% -11.054%] (p = 0.00 < 0.05)
                        thrpt:  [+12.428% +12.778% +13.077%]
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  3 (3.00%) high mild
  6 (6.00%) high severe
librustdoc/sidebar      time:   [21.856 µs 21.875 µs 21.896 µs]
                        thrpt:  [56.359 MiB/s 56.413 MiB/s 56.463 MiB/s]
                 change:
                        time:   [-17.757% -16.668% -15.597%] (p = 0.00 < 0.05)
                        thrpt:  [+18.479% +20.002% +21.591%]
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high severe
librustdoc/source       time:   [7.5811 µs 7.5986 µs 7.6122 µs]
                        thrpt:  [96.844 MiB/s 97.017 MiB/s 97.241 MiB/s]
                 change:
                        time:   [-17.455% -17.335% -17.224%] (p = 0.00 < 0.05)
                        thrpt:  [+20.807% +20.970% +21.146%]
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  3 (3.00%) low severe
  5 (5.00%) low mild
  2 (2.00%) high mild
librustdoc/type_layout_size
                        time:   [4.9870 µs 4.9936 µs 5.0009 µs]
                        thrpt:  [54.159 MiB/s 54.238 MiB/s 54.310 MiB/s]
                 change:
                        time:   [-9.9132% -9.5887% -9.2205%] (p = 0.00 < 0.05)
                        thrpt:  [+10.157% +10.606% +11.004%]
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  3 (3.00%) high mild
  6 (6.00%) high severe
librustdoc/type_layout  time:   [17.952 µs 18.117 µs 18.302 µs]
                        thrpt:  [147.10 MiB/s 148.60 MiB/s 149.97 MiB/s]
                 change:
                        time:   [-12.943% -12.385% -11.870%] (p = 0.00 < 0.05)
                        thrpt:  [+13.468% +14.135% +14.867%]
                        Performance has improved.
Found 17 outliers among 100 measurements (17.00%)
  1 (1.00%) low severe
  15 (15.00%) low mild
  1 (1.00%) high mild
```

</details>

### CODE REVIEW

**Comment by Kijewski** [comment]
Thank you very much for working on this change, really impressive results! I tried out winnow in a toy project (a simple e-mail address parser), and found it much more ergonomic to use than nom. Same in your code, it is more readable now than it was before.

**Review by Kijewski — DISMISSED** [review]

File: rinja_parser/src/expr.rs [hunk 1]
```diff
@@ -1,18 +1,19 @@
...
-use nom::sequence::{pair, preceded, terminated, tuple};
+use winnow::branch::alt;
+use winnow::bytes::{one_of, tag, take_till0};
+use winnow::character::digit1;
+use winnow::combinator::{cut_err, fail, not, opt, peek};
+use winnow::error::ErrorKind;
+use winnow::error::ParseError as _;
+use winnow::multi::{fold_many0, many0, separated0, separated1};
+use winnow::sequence::{preceded, terminated};
+use winnow::Parser;
```
Comment (Kijewski) [inline rinja_parser/src/expr.rs:12]: We use nightly to format our code, I guess that's why it isn't formatted as expected: `cargo +nightly fmt`

**Review by Kijewski — DISMISSED** [review]

**Review by Kijewski — APPROVED** [review]

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
This is awesome, thanks a lot @epage! :)

**Comment by epage** [comment]
Note: we're losing some of the performance gains in upgrading to 0.4.  See #210.

**Pull request merged**